### PR TITLE
API provides create action for landlord

### PR DESCRIPTION
**Pull request by Odeane**
### User Story
```
As a API,
In order for landlord to post their ads
I would like to provide a rental space create action.
```
## Changes proposed in this pull request:
* Landlord can create listings

## What I have learned working on this feature:
* How to use geocoder

## Packages/Gems added
- Geocoder
- Devise_token_auth

### CODE REVIEW

File: app/controllers/api/v1/listings_controller.rb [hunk 1]
```diff
@@ -1,14 +1,35 @@
+# frozen_string_literal: true
+
 class Api::V1::ListingsController < ApplicationController
+  before_action :authenticate_user!, only: [:create]
+
   def index
-   listings = Listing.all  
-   render json: {listings: listings}, each_serializer: ListingIndexSerializer
+    listings = Listing.all
+    render json: { listings: listings }, each_serializer: ListingIndexSerializer
```
Comment (viktorsommar) [inline app/controllers/api/v1/listings_controller.rb:8]: what happened to the identation?

File: app/controllers/application_controller.rb [hunk 2]
```diff
@@ -1,2 +1,14 @@
 class ApplicationController < ActionController::API
+  include DeviseTokenAuth::Concerns::SetUserByToken
+
+	private
```
Comment (viktorsommar) [inline app/controllers/application_controller.rb]: what happened with formatting here?

File: app/models/user.rb [hunk 3]
```diff
@@ -0,0 +1,11 @@
+# frozen_string_literal: true
+
+class User < ActiveRecord::Base
+  extend Devise::Models
+  devise :database_authenticatable, :registerable,
+         :recoverable, :rememberable, :trackable, :validatable
+  include DeviseTokenAuth::Concerns::User
+
+  has_many :listings, foreign_key: "landlord_id", class_name: "Listing"
```
Comment (viktorsommar) [inline app/models/user.rb:9]: what is the landlord_id?

File: config/routes.rb [hunk 4]
```diff
@@ -1,8 +1,9 @@
 Rails.application.routes.draw do
+  mount_devise_token_auth_for 'User', at: 'auth'
```
Comment (viktorsommar) [inline config/routes.rb]: Should this not be "api/v1/auth"?

File: spec/factories/listings.rb [hunk 5]
```diff
@@ -6,5 +6,6 @@
     description { 'Gated park with night lamp' }
     price { 200 }
     address {'Lövåsvägen 21 16733 Bromma'}
+    association :landlord, factory: :user
```
Comment (viktorsommar) [inline spec/factories/listings.rb:9]: What is the landlord association?

File: spec/factories/user.rb [hunk 6]
```diff
@@ -0,0 +1,6 @@
+FactoryBot.define do
+  factory :user do
+    email { "user#{rand(1...9999)}@mail.com" }
```
Comment (viktorsommar) [inline spec/factories/user.rb:3]: Is there a specific reason why we make the email like this?

File: app/controllers/application_controller.rb [hunk 7]
```diff
@@ -1,2 +1,14 @@
...
+	
+  def render_error_message(errors)
```
Comment (oliverochman) [inline app/controllers/application_controller.rb:8]: please correct the indentation in this file

**Review by oliverochman — CHANGES_REQUESTED** [review]

File: spec/requests/api/v1/listings/create_spec.rb [hunk 8]
```diff
@@ -0,0 +1,119 @@
+# frozen_string_literal: true
+
+RSpec.describe 'POST /api/v1/listings', type: :request do
+  let!(:landlord) { create(:user) }
```
Comment (viktorsommar) [inline spec/requests/api/v1/listings/create_spec.rb:4]: all users should be landlords and be able to rent in the same account

**Review by oliverochman — APPROVED** [review]

**Review by viktorsommar — CHANGES_REQUESTED** [review]

File: app/controllers/api/v1/listings_controller.rb (hunk 1 above)
Comment (Odeane) [reply app/controllers/api/v1/listings_controller.rb:8]: What indention sir???

File: app/models/user.rb (hunk 3 above)
Comment (Odeane) [reply app/models/user.rb:9]: association

File: app/controllers/api/v1/listings_controller.rb (hunk 1 above)
Comment (viktorsommar) [reply app/controllers/api/v1/listings_controller.rb:8]: It's been fixed now

File: spec/factories/user.rb (hunk 6 above)
Comment (Cma-B) [reply spec/factories/user.rb:3]: cause we need to have random email

File: spec/factories/listings.rb (hunk 5 above)
Comment (Cma-B) [reply spec/factories/listings.rb:9]: each listing should have a landlord

File: spec/requests/api/v1/listings/create_spec.rb (hunk 8 above)
Comment (Cma-B) [reply spec/requests/api/v1/listings/create_spec.rb:4]: true! in the test we made a user and called it landlord as we have a foreign key of landlord_id in listing data-base

File: spec/factories/listings.rb (hunk 5 above)
Comment (viktorsommar) [reply spec/factories/listings.rb:9]: But we should not have a specific landlord role, it should be a user or a subscriber

**Review by Alex-Bellamy — CHANGES_REQUESTED** [review]
there is only one type of user, think you have split it out to landlord in error

**Comment by Cma-B** [comment]
landlord as user that can create the listing is just a variable that we have used to make the flow more descriptive! so we have just one type of  user no worries!

**Pull request merged**